src/lsm/tree.zig

```diff
@@ -266,6 +272,88 @@ pub fn TreeType(comptime TreeTable: type, comptime Storage: type, comptime tree_
             tree.table_mutable.remove(value);
         }
 
+        pub fn update_key_range(tree: *Tree, key: Key) void {
+            assert((tree.key_min == null and tree.key_max == null) or
+                (tree.key_min != null and tree.key_max != null));
+            if (tree.key_min == null or compare_keys(key, tree.key_min.?) == .lt) {
+                tree.key_min = key;
+            }
+            if (tree.key_max == null or compare_keys(key, tree.key_max.?) == .gt) {
+                tree.key_max = key;
+            }
+            assert(tree.key_min != null and tree.key_max != null);
+        }
+
+        /// Returns True if the given key may be present in the Tree/Manifest, False if the key is
+        /// guaranteed to not be present.
+        ///
+        /// Performs the following steps based on the provided location:
+        /// *  tree -> Checks whether the key exists within the Tree's key range.
+        /// *  manifest -> Checks whether the key exists within the ManifestLevels' key range.
+        ///    Then, for each ManifestLevel that could contain the key, it attempts to fetch the
+        ///    index block, filter block & data block for the tables that could contain the key
+        ///    synchronously from the Grid cache. If any of these blocks aren't present in the cache,
+        ///    it returns True since the existence of the key cannot be determined.
+        pub fn key_may_exist(tree: *Tree, key: Key, location: union(enum) { tree, manifest }) bool {
+            assert((tree.key_min == null and tree.key_max == null) or
+                (tree.key_min != null and tree.key_max != null));
+            switch (location) {
+                .tree => {
```

Comment:
Here too, better to have two explicit functions (`key_may_exist_in_tree/manifest`), than a function that branches dynamically.